python/hail/matrixtable.py

```diff
@@ -391,16 +386,21 @@ def __init__(self, jvds):
         assert isinstance(self._row_type, tstruct), self._row_type
         assert isinstance(self._entry_type, tstruct), self._entry_type
 
-        self._globals = construct_expr(TopLevelReference('global', self._global_indices), self._global_type,
-                                       indices=self._global_indices)
-        self._rvrow = construct_expr(TopLevelReference('va', self._row_indices), HailType._from_java(jvds.rvRowType()),
-                                   indices=self._row_indices)
-        self._row = hail.struct(
-            **{k: self._rvrow[k] for k in self._row_type.keys()})
-        self._col = construct_expr(TopLevelReference('sa', self._col_indices), self._col_type,
-                                   indices=self._col_indices)
-        self._entry = construct_expr(TopLevelReference('g', self._entry_indices), self._entry_type,
-                                     indices=self._entry_indices)
+        self._globals = construct_reference('global', self._global_type,
+                                            indices=self._global_indices)
+        self._rvrow = construct_reference('va',
+                                          HailType._from_java(jvds.rvRowType()),
+                                          indices=self._row_indices)
+        self._row = hail.struct(**{k: self._rvrow[k] for k in self._row_type.keys()})
```

Comment:
let's emit a SelectFields instead of a MakeStruct here?

Reply:
actually, Cotton wrote an optimizer rule around this. But it is probably still going to generate bigger IRs.

Reply:
good call.

Reply:
actually, I'm reverting this change for now because there's an amount of logic scattered throughout that relies on this being the way it is to do field checking/getting.

I'll follow up this PR to use select instead of the splat here and in some other places (like _row_key, etc.)